// react-auth-kit/dist/utils/hooks.js
"use strict";
'use client';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useInterval = useInterval;
exports.useIntervalRefresh = useIntervalRefresh;
exports.useIsomorphicLayoutEffect = void 0;
exports.useNavigateTo = useNavigateTo;
exports.useTryNavigateTo = useTryNavigateTo;
var _react = require("react");
var _AuthContext = require("../AuthContext");
var _error = require("../error");
var useIsomorphicLayoutEffect = exports.useIsomorphicLayoutEffect = typeof window !== 'undefined' ? _react.useLayoutEffect : _react.useEffect;
function useInterval(callback, delay) {
  var savedCallback = (0, _react.useRef)(callback);
  useIsomorphicLayoutEffect(function () {
    savedCallback.current = callback;
  }, [callback]);
  (0, _react.useEffect)(function () {
    if (delay === null) {
      return;
    }
    var id = setInterval(function () {
      savedCallback.current();
    }, delay);
    return function () {
      clearInterval(id);
    };
  }, [delay]);
}
function useIntervalRefresh(callback, store, interval) {
  var intervalId;
  store.subscribe(function (state) {
    if (state.isSignIn) {
      if (intervalId) {
        clearInterval(intervalId);
      }
      intervalId = setInterval(function () {
        callback();
      }, interval);
    } else {
      if (intervalId) {
        clearInterval(intervalId);
        intervalId = undefined;
      }
    }
  });
}
function useNavigateTo() {
  var router = (0, _AuthContext.useReactAuthKitRouter)();
  var navigate = router ? router.useNavigate() : null;
  return function (to) {
    if (router && navigate) {
      navigate({
        to: to
      });
    } else {
      throw new _error.AuthKitError('Router Plugin is not implemented in the AuthProvider. Please use the router prop of AuthProvider and Router plugin to use this feature');
    }
  };
}
function useTryNavigateTo() {
  var navigateTo = useNavigateTo();
  return function (to) {
    if (to) {
      navigateTo(to);
    }
  };
}
