// react-auth-kit/dist/store/TokenStore.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _rxjs = require("rxjs");
var _deepEqual = _interopRequireDefault(require("deep-equal"));
var _tryOrDefault = _interopRequireDefault(require("../utils/tryOrDefault"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _classCallCheck(a, n) { if (!(a instanceof n)) throw new TypeError("Cannot call a class as a function"); }
function _defineProperties(e, r) { for (var t = 0; t < r.length; t++) { var o = r[t]; o.enumerable = o.enumerable || !1, o.configurable = !0, "value" in o && (o.writable = !0), Object.defineProperty(e, _toPropertyKey(o.key), o); } }
function _createClass(e, r, t) { return r && _defineProperties(e.prototype, r), t && _defineProperties(e, t), Object.defineProperty(e, "prototype", { writable: !1 }), e; }
function _defineProperty(e, r, t) { return (r = _toPropertyKey(r)) in e ? Object.defineProperty(e, r, { value: t, enumerable: !0, configurable: !0, writable: !0 }) : e[r] = t, e; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
var TokenStore = function () {
  function TokenStore(isUsingRefreshToken, storage, storageNamingStrategy, token, debug) {
    var _this = this;
    _classCallCheck(this, TokenStore);
    _defineProperty(this, "subscribe", function (next, error, complete) {
      _this.authSubject.subscribe({
        next: next,
        error: error,
        complete: complete
      });
    });
    _defineProperty(this, "set", function (data) {
      _this.log("Set Function is called with", data);
      _this.log("Set Function Old Data", _this.value);
      var obj = _objectSpread({}, _this.value);
      if (data.userState !== undefined) {
        obj.userState = data.userState;
      }
      if (data.auth) {
        try {
          var exp = _this.token.getExpiresAt(data.auth.token);
          if (exp > new Date()) {
            obj = _objectSpread(_objectSpread({}, obj), {}, {
              auth: {
                'token': data.auth.token,
                'type': data.auth.type,
                'expiresAt': exp
              },
              isSignIn: true
            });
          } else {
            obj = _objectSpread(_objectSpread({}, obj), {}, {
              auth: null,
              isSignIn: false,
              userState: null
            });
          }
        } catch (_e) {
          console.warn(_e);
          obj = _objectSpread(_objectSpread({}, obj), {}, {
            auth: null,
            isSignIn: false,
            userState: null
          });
        }
      } else if (data.auth === null) {
        obj = _objectSpread(_objectSpread({}, obj), {}, {
          auth: null,
          isSignIn: false,
          userState: null
        });
      }
      if (_this.isUsingRefreshToken) {
        if (obj.auth === null) {
          obj = _objectSpread(_objectSpread({}, obj), {}, {
            refresh: null
          });
        } else if (data.refresh) {
          try {
            var refreshExpireTime = _this.token.getExpiresAt(data.refresh);
            if (refreshExpireTime > new Date()) {
              obj = _objectSpread(_objectSpread({}, obj), {}, {
                refresh: {
                  'token': data.refresh,
                  'expiresAt': refreshExpireTime
                },
                isUsingRefreshToken: true
              });
            } else {
              obj = _objectSpread(_objectSpread({}, obj), {}, {
                refresh: null,
                auth: null,
                isSignIn: false,
                userState: null
              });
            }
          } catch (e) {
            _this.log("Error Occurred in setting the refresh token", e);
            console.warn(e);
            obj = _objectSpread(_objectSpread({}, obj), {}, {
              refresh: null,
              auth: null,
              isSignIn: false,
              userState: null
            });
          }
        } else if (data.refresh === null) {
          obj = _objectSpread(_objectSpread({}, obj), {}, {
            refresh: null
          });
        }
      }
      _this.log("Set Function New Data", obj);
      if (!(0, _deepEqual.default)(_this.value, obj)) {
        _this.log('Updating the value in the Set Function', _this.value, obj);
        _this.authValue = obj;
        _this.authSubject.next(obj);
      }
    });
    _defineProperty(this, "initialToken_", function () {
      try {
        var authToken = _tryOrDefault.default.method(_this.storage.get, _this.storage, null, _this.storageNamingStrategy.getAuthStorageName());
        var authTokenType = _tryOrDefault.default.method(_this.storage.get, _this.storage, null, _this.storageNamingStrategy.getAuthTypeStorageName());
        var stateCookie = _tryOrDefault.default.method(_this.storage.get, _this.storage, null, _this.storageNamingStrategy.getStateStorageName());
        var refreshToken = _this.isUsingRefreshToken ? _tryOrDefault.default.method(_this.storage.get, _this.storage, null, _this.storageNamingStrategy.getRefreshTokenStorageName()) : null;
        return _this.checkTokenExist_(authToken, authTokenType, stateCookie, refreshToken);
      } catch (e) {
        console.error(e);
        _this.log("Error Occurred in initialToken_()");
        return {
          auth: null,
          refresh: null,
          userState: null,
          isUsingRefreshToken: _this.isUsingRefreshToken,
          isSignIn: false
        };
      }
    });
    _defineProperty(this, "checkTokenExist_", function (authToken, authTokenType, stateCookie, refreshToken) {
      _this.log('checkTokenExist_ is called');
      _this.log("Params: authToken: ".concat(authToken, ", authTokenType: ").concat(authTokenType, ",\n      stateCookie: ").concat(stateCookie, ", refreshToken: ").concat(refreshToken));
      try {
        var refresh;
        if (_this.isUsingRefreshToken && !!refreshToken) {
          _this.log("checkTokenExist - isUsingRefreshToken\n          = ".concat(_this.isUsingRefreshToken, " refreshToken - ").concat(refreshToken));
          var refreshTokenExpiresAt = _this.token.getExpiresAt(refreshToken);
          if (_this.token.getExpiresAt(refreshToken) < new Date()) {
            _this.log("checkTokenExist - refresh token is expired\n            ".concat(refreshTokenExpiresAt, " ").concat(new Date()));
            refresh = null;
          } else {
            _this.log("checkTokenExist - new refresh token is assigned\n            ".concat(refreshToken));
            refresh = {
              token: refreshToken,
              expiresAt: refreshTokenExpiresAt
            };
          }
        } else {
          _this.log("checkTokenExist - Refresh Token is invalid or not using\n           refresh feature ".concat(_this.isUsingRefreshToken, " ").concat(refreshToken));
          refresh = null;
        }
        if (_this.isUsingRefreshToken && !refresh) {
          _this.log("checkTokenExist - Removing Refresh Token");
          _this.removeAllToken();
          return {
            auth: null,
            refresh: null,
            userState: null,
            isUsingRefreshToken: _this.isUsingRefreshToken,
            isSignIn: false
          };
        }
        var auth;
        var authState;
        if (!!authToken && !!authTokenType && !!stateCookie) {
          _this.log("checkTokenExist - authToken, authTokenType, stateCookie exists");
          try {
            var expiresAt = _this.token.getExpiresAt(authToken);
            if (expiresAt < new Date()) {
              _this.log("checkTokenExist - auth token is expired\n              ".concat(expiresAt, " ").concat(new Date()));
              auth = null;
              authState = null;
            } else {
              try {
                authState = JSON.parse(stateCookie.replaceAll('\\', ''));
                auth = {
                  token: authToken,
                  type: authTokenType,
                  expiresAt: expiresAt
                };
              } catch (_e) {
                _this.log('state cookie JSON parsing failed ${err}');
                auth = null;
                authState = null;
              }
            }
          } catch (e) {
            _this.log("checkTokenExist - auth token or auth state is invalid\n            ".concat(authToken, " ").concat(stateCookie));
            _this.log("Error Occurred: ".concat(e));
            auth = null;
            authState = null;
          }
        } else {
          _this.log("checkTokenExist " + "- authToken, authTokenType, stateCookie doesn't exists");
          auth = null;
          authState = null;
        }
        if (_this.isUsingRefreshToken && refresh) {
          if (!!auth && !!authState) {
            _this.log('checkTokenExist - Returning auth and refresh');
            _this.log({
              auth: auth,
              refresh: refresh,
              userState: authState,
              isUsingRefreshToken: _this.isUsingRefreshToken,
              isSignIn: true
            });
            return {
              auth: auth,
              refresh: refresh,
              userState: authState,
              isUsingRefreshToken: _this.isUsingRefreshToken,
              isSignIn: true
            };
          }
          _this.log('checkTokenExist - Removing Auth Token');
          _this.removeAuth();
          _this.log({
            auth: null,
            refresh: refresh,
            userState: null,
            isUsingRefreshToken: _this.isUsingRefreshToken,
            isSignIn: false
          });
          return {
            auth: null,
            refresh: refresh,
            userState: null,
            isUsingRefreshToken: _this.isUsingRefreshToken,
            isSignIn: false
          };
        } else if (!_this.isUsingRefreshToken && !!auth && !!authState) {
          _this.log('checkTokenExist - Returning auth');
          _this.log({
            auth: auth,
            refresh: null,
            userState: authState,
            isUsingRefreshToken: _this.isUsingRefreshToken,
            isSignIn: true
          });
          return {
            auth: auth,
            refresh: null,
            userState: authState,
            isUsingRefreshToken: _this.isUsingRefreshToken,
            isSignIn: true
          };
        }
        {
          _this.log('checkTokenExist- removing all tokens. Returning null');
          _this.removeAllToken();
          return {
            auth: null,
            refresh: null,
            userState: null,
            isUsingRefreshToken: _this.isUsingRefreshToken,
            isSignIn: false
          };
        }
      } catch (e) {
        _this.log("Error Occurred in checkTokenExist_()", e);
        _this.removeAllToken();
        return {
          auth: null,
          refresh: null,
          userState: null,
          isUsingRefreshToken: _this.isUsingRefreshToken,
          isSignIn: false
        };
      }
    });
    _defineProperty(this, "syncTokens", function (authState) {
      if (authState.auth) {
        _this.setAuthToken(authState.auth.token, authState.auth.type, authState.userState);
      } else {
        _this.removeAuth();
      }
      if (!!authState.refresh && _this.isUsingRefreshToken) {
        _this.setRefreshToken(authState.refresh.token);
      } else {
        _this.removeRefresh();
      }
    });
    _defineProperty(this, "setAuthToken", function (authToken, authTokenType, authState) {
      var expiresAt = _this.token.getExpiresAt(authToken);
      _this.storage.set(_this.storageNamingStrategy.getAuthStorageName(), authToken, expiresAt);
      _this.storage.set(_this.storageNamingStrategy.getAuthTypeStorageName(), authTokenType, expiresAt);
      if (authState) {
        _this.storage.set(_this.storageNamingStrategy.getStateStorageName(), JSON.stringify(authState), expiresAt);
      }
    });
    _defineProperty(this, "setRefreshToken", function (refreshToken) {
      if (_this.isUsingRefreshToken && !!refreshToken) {
        var refreshTokenExpiresAt = _this.token.getExpiresAt(refreshToken);
        _this.storage.set(_this.storageNamingStrategy.getRefreshTokenStorageName(), refreshToken, refreshTokenExpiresAt);
      }
    });
    _defineProperty(this, "removeAllToken", function () {
      _this.storage.remove(_this.storageNamingStrategy.getAuthStorageName());
      _this.storage.remove(_this.storageNamingStrategy.getAuthTypeStorageName());
      _this.storage.remove(_this.storageNamingStrategy.getStateStorageName());
      if (_this.isUsingRefreshToken) {
        _this.storage.remove(_this.storageNamingStrategy.getRefreshTokenStorageName());
      }
    });
    _defineProperty(this, "removeAuth", function () {
      _this.storage.remove(_this.storageNamingStrategy.getAuthStorageName());
      _this.storage.remove(_this.storageNamingStrategy.getAuthTypeStorageName());
      _this.storage.remove(_this.storageNamingStrategy.getStateStorageName());
    });
    _defineProperty(this, "removeRefresh", function () {
      if (_this.isUsingRefreshToken) {
        _this.storage.remove(_this.storageNamingStrategy.getRefreshTokenStorageName());
      }
    });
    _defineProperty(this, "log", function (msg) {
      if (_this.debug) {
        for (var _len = arguments.length, optionalParams = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          optionalParams[_key - 1] = arguments[_key];
        }
        console.log("React Auth Kit - Message ".concat(JSON.stringify(msg), " [ ").concat(JSON.stringify(optionalParams), " ]"));
      }
    });
    this.storage = storage;
    this.storageNamingStrategy = storageNamingStrategy;
    this.token = token;
    this.debug = debug;
    this.isUsingRefreshToken = isUsingRefreshToken;
    this.authValue = this.initialToken_();
    this.authSubject = new _rxjs.BehaviorSubject(this.authValue);
    this.log("Initial Value", this.authValue);
    this.authSubject.pipe((0, _rxjs.skip)(1)).subscribe({
      next: this.syncTokens
    });
  }
  return _createClass(TokenStore, [{
    key: "value",
    get: function get() {
      return this.authSubject.value;
    }
  }]);
}();
var _default = exports.default = TokenStore;
